File: backend/app/pipeline/matching_engine.py
"""
Matching engine for calculating job-resume compatibility scores.
"""
from typing import List, Dict, Any, Tuple
from dataclasses import dataclass
import structlog

from .skill_extraction import calculate_skill_match
from .embeddings import calculate_cosine_similarity

logger = structlog.get_logger(__name__)


@dataclass
class MatchResult:
    """Result of a single match calculation."""
    resume_id: str
    job_id: str
    student_id: str
    recruiter_id: str
    overall_score: float
    skill_score: float
    experience_score: float
    semantic_score: float
    matched_skills: List[str]
    missing_skills: List[str]
    explanation: Dict[str, Any]


class MatchingEngine:
    """Engine for calculating match scores between resumes and jobs."""

    @staticmethod
    async def calculate_match_score(
        resume_data: Dict[str, Any],
        job_data: Dict[str, Any]
    ) -> MatchResult:
        """
        Calculate comprehensive match score between a resume and job.

        Args:
            resume_data: Resume information including skills, experience, embedding
            job_data: Job information including requirements, embedding

        Returns:
            MatchResult with detailed scoring
        """
        logger.info(
            "Calculating match score",
            resume_id=resume_data.get("id"),
            job_id=job_data.get("id")
        )

        # Extract data
        resume_skills = resume_data.get("extracted_skills", [])
        job_skills = job_data.get("required_skills", [])
        resume_embedding = resume_data.get("embedding_vector", [])
        job_embedding = job_data.get("embedding_vector", [])
        resume_experience = resume_data.get("experience", [])
        job_level = job_data.get("experience_level", "entry")

        # Calculate skill score
        skill_match = calculate_skill_match(resume_skills, job_skills)
        skill_score = skill_match["match_score"]

        # Calculate semantic score (embedding similarity)
        semantic_score = 0.0
        if resume_embedding and job_embedding:
            semantic_score = await calculate_cosine_similarity(
                resume_embedding, job_embedding
            )

        # Calculate experience score
        experience_score = _calculate_experience_score(resume_experience, job_level)

        # Calculate overall score (weighted combination)
        overall_score = _calculate_overall_score(
            skill_score, semantic_score, experience_score
        )

        # Generate explanation
        explanation = _generate_match_explanation(
            skill_match, semantic_score, experience_score, overall_score
        )

        return MatchResult(
            resume_id=resume_data["id"],
            job_id=job_data["id"],
            student_id=resume_data["user_id"],
            recruiter_id=job_data["recruiter_id"],
            overall_score=overall_score,
            skill_score=skill_score,
            experience_score=experience_score,
            semantic_score=semantic_score,
            matched_skills=skill_match["matched"],
            missing_skills=skill_match["missing"],
            explanation=explanation,
        )

    @staticmethod
    async def batch_calculate_matches(
        resumes: List[Dict[str, Any]],
        jobs: List[Dict[str, Any]]
    ) -> List[MatchResult]:
        """
        Calculate match scores for multiple resume-job pairs.

        Args:
            resumes: List of resume data
            jobs: List of job data

        Returns:
            List of match results
        """
        logger.info(
            "Calculating batch matches",
            resume_count=len(resumes),
            job_count=len(jobs)
        )

        results = []

        for resume in resumes:
            for job in jobs:
                try:
                    result = await MatchingEngine.calculate_match_score(resume, job)
                    results.append(result)
                except Exception as e:
                    logger.error(
                        "Failed to calculate match",
                        resume_id=resume.get("id"),
                        job_id=job.get("id"),
                        error=str(e)
                    )

        # Sort by overall score (descending)
        results.sort(key=lambda x: x.overall_score, reverse=True)

        logger.info("Batch matching completed", total_matches=len(results))
        return results


def _calculate_experience_score(resume_experience: List[Dict], job_level: str) -> float:
    """
    Calculate experience compatibility score.

    Args:
        resume_experience: List of experience entries
        job_level: Required experience level

    Returns:
        Experience score (0-1)
    """
    if not resume_experience:
        return 0.0

    # Map job levels to years of experience
    level_requirements = {
        "internship": 0,
        "entry": 0,
        "mid": 2,
        "senior": 5,
    }

    required_years = level_requirements.get(job_level, 0)

    # Calculate total years of experience
    total_years = 0
    for exp in resume_experience:
        # Simple calculation - can be enhanced with date parsing
        duration = exp.get("duration_years", 0)
        total_years += duration

    # Score based on meeting requirements
    if total_years >= required_years:
        if total_years >= required_years + 3:  # Overqualified
            return 0.9
        return 1.0
    elif total_years >= required_years * 0.5:  # Partial match
        return 0.6
    else:
        return 0.2


def _calculate_overall_score(
    skill_score: float,
    semantic_score: float,
    experience_score: float
) -> float:
    """
    Calculate weighted overall score.

    Weights:
    - Skills: 40%
    - Semantic similarity: 35%
    - Experience: 25%
    """
    return (
        skill_score * 0.4 +
        semantic_score * 0.35 +
        experience_score * 0.25
    )


def _generate_match_explanation(
    skill_match: Dict,
    semantic_score: float,
    experience_score: float,
    overall_score: float
) -> Dict[str, Any]:
    """
    Generate human-readable explanation for the match.
    """
    # Determine match quality
    if overall_score >= 0.8:
        quality = "excellent"
        summary = "Excellent match! Strong alignment across all criteria."
    elif overall_score >= 0.6:
        quality = "good"
        summary = "Good match with solid qualifications."
    elif overall_score >= 0.4:
        quality = "moderate"
        summary = "Moderate match with some alignment."
    else:
        quality = "poor"
        summary = "Limited match, significant gaps present."

    # Generate strengths
    strengths = []
    if skill_match["match_score"] >= 0.7:
        strengths.append("Strong skill alignment")
    if semantic_score >= 0.7:
        strengths.append("High semantic similarity")
    if experience_score >= 0.8:
        strengths.append("Excellent experience match")

    # Generate gaps
    gaps = []
    if skill_match["missing"]:
        gaps.append(f"Missing {len(skill_match['missing'])} required skills")
    if semantic_score < 0.5:
        gaps.append("Limited semantic alignment")
    if experience_score < 0.5:
        gaps.append("Experience level mismatch")

    # Generate recommendations
    recommendations = []
    if skill_match["missing"]:
        recommendations.append(f"Consider learning: {', '.join(skill_match['missing'][:3])}")
    if experience_score < 0.7:
        recommendations.append("Gain more relevant experience")
    if semantic_score < 0.6:
        recommendations.append("Tailor resume to highlight relevant keywords")

    return {
        "summary": summary,
        "quality": quality,
        "strengths": strengths,
        "gaps": gaps,
        "recommendations": recommendations,
        "skill_breakdown": {
            "matched": skill_match["matched"],
            "missing": skill_match["missing"],
            "additional": skill_match["additional"],
        },
        "score_breakdown": {
            "skills": skill_match["match_score"],
            "semantic": semantic_score,
            "experience": experience_score,
            "overall": overall_score,
        },
    }

File: src/services/ml/matchingEngine.ts
/**
 * Matching Engine
 * Core ML service for matching resumes to jobs
 */

import { logger } from '@/utils/logger';
import { AppConfig } from '@/config/app.config';
import { cosineSimilarity } from './embeddings';
import { calculateSkillOverlap } from './skillExtraction';
import type { Resume, Job, Match, MatchExplanation } from '@/types/models';
import { generateId } from '@/utils/helpers';

export interface MatchScore {
  overallScore: number;
  skillScore: number;
  experienceScore: number;
  semanticScore: number;
}

export interface MatchResult extends MatchScore {
  resumeId: string;
  jobId: string;
  studentId: string;
  recruiterId: string;
  matchedSkills: string[];
  missingSkills: string[];
  explanation: MatchExplanation;
}

/**
 * Match a single resume to a single job
 */
export function matchResumeToJob(resume: Resume, job: Job): MatchResult {
  logger.info('Matching resume to job', {
    resumeId: resume.id,
    jobId: job.id,
  });

  // Calculate individual scores
  const skillScore = calculateSkillScore(resume, job);
  const experienceScore = calculateExperienceScore(resume, job);
  const semanticScore = calculateSemanticScore(resume, job);

  // Calculate weighted overall score
  const overallScore =
    skillScore * AppConfig.ml.skillWeightage +
    experienceScore * AppConfig.ml.experienceWeightage +
    semanticScore * AppConfig.ml.semanticWeightage;

  // Identify matched and missing skills
  const { matched, onlyInSecond: missing } = calculateSkillOverlap(
    resume.extractedSkills,
    job.requiredSkills
  );

  // Generate explanation
  const explanation = generateMatchExplanation(
    resume,
    job,
    { overallScore, skillScore, experienceScore, semanticScore },
    matched,
    missing
  );

  logger.info('Match completed', {
    resumeId: resume.id,
    jobId: job.id,
    overallScore: overallScore.toFixed(3),
  });

  return {
    resumeId: resume.id,
    jobId: job.id,
    studentId: resume.userId,
    recruiterId: job.recruiterId,
    overallScore,
    skillScore,
    experienceScore,
    semanticScore,
    matchedSkills: matched,
    missingSkills: missing,
    explanation,
  };
}

/**
 * Match a resume to multiple jobs and return top K matches
 */
export function matchResumeToJobs(
  resume: Resume,
  jobs: Job[],
  topK: number = AppConfig.ml.topK
): MatchResult[] {
  logger.info('Matching resume to multiple jobs', {
    resumeId: resume.id,
    jobCount: jobs.length,
    topK,
  });

  const matches = jobs
    .filter(job => job.status === 'active')
    .map(job => matchResumeToJob(resume, job))
    .sort((a, b) => b.overallScore - a.overallScore)
    .slice(0, topK);

  logger.info('Multi-job matching completed', {
    matchCount: matches.length,
  });

  return matches;
}

/**
 * Match a job to multiple resumes and return top K candidates
 */
export function matchJobToResumes(
  job: Job,
  resumes: Resume[],
  topK: number = AppConfig.ml.topK
): MatchResult[] {
  logger.info('Matching job to multiple resumes', {
    jobId: job.id,
    resumeCount: resumes.length,
    topK,
  });

  const matches = resumes
    .filter(resume => resume.status === 'completed')
    .map(resume => matchResumeToJob(resume, job))
    .sort((a, b) => b.overallScore - a.overallScore)
    .slice(0, topK);

  logger.info('Multi-resume matching completed', {
    matchCount: matches.length,
  });

  return matches;
}

/**
 * Calculate skill match score
 */
function calculateSkillScore(resume: Resume, job: Job): number {
  const resumeSkills = new Set(resume.extractedSkills.map(s => s.toLowerCase()));
  const requiredSkills = job.requiredSkills;
  const preferredSkills = job.preferredSkills || [];

  let score = 0;
  let totalWeight = 0;

  // Required skills (higher weight)
  for (const skill of requiredSkills) {
    totalWeight += 1.0;
    if (resumeSkills.has(skill.toLowerCase())) {
      score += 1.0;
    }
  }

  // Preferred skills (lower weight)
  for (const skill of preferredSkills) {
    totalWeight += 0.5;
    if (resumeSkills.has(skill.toLowerCase())) {
      score += 0.5;
    }
  }

  return totalWeight > 0 ? score / totalWeight : 0;
}

/**
 * Calculate experience level match score
 */
function calculateExperienceScore(resume: Resume, job: Job): number {
  const experienceYears = estimateExperienceYears(resume);
  const requiredLevel = job.experienceLevel;

  const levelMap: Record<string, { min: number; max: number }> = {
    internship: { min: 0, max: 1 },
    entry: { min: 0, max: 2 },
    mid: { min: 2, max: 5 },
    senior: { min: 5, max: 100 },
  };

  const required = levelMap[requiredLevel];
  if (!required) return 0.5; // Default score if level unknown

  // Perfect match if within range
  if (experienceYears >= required.min && experienceYears <= required.max) {
    return 1.0;
  }

  // Partial match if close
  if (experienceYears < required.min) {
    const deficit = required.min - experienceYears;
    return Math.max(0, 1.0 - deficit * 0.2); // -20% per year deficit
  } else {
    const excess = experienceYears - required.max;
    return Math.max(0.5, 1.0 - excess * 0.1); // -10% per year excess, min 50%
  }
}

/**
 * Estimate years of experience from resume
 */
function estimateExperienceYears(resume: Resume): number {
  let totalYears = 0;

  for (const exp of resume.experience) {
    if (exp.startDate && exp.endDate) {
      const start = new Date(exp.startDate);
      const end = exp.endDate.toLowerCase() === 'present' 
        ? new Date() 
        : new Date(exp.endDate);
      
      const years = (end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24 * 365);
      totalYears += years;
    } else {
      // Estimate 1 year if dates not provided
      totalYears += 1;
    }
  }

  return totalYears;
}

/**
 * Calculate semantic similarity score using embeddings
 */
function calculateSemanticScore(resume: Resume, job: Job): number {
  if (!resume.embeddingVector || !job.embeddingVector) {
    logger.warn('Missing embedding vectors, using fallback score');
    return 0.5;
  }

  const similarity = cosineSimilarity(resume.embeddingVector, job.embeddingVector);
  
  // Normalize to 0-1 range (cosine similarity is already -1 to 1)
  return (similarity + 1) / 2;
}

/**
 * Generate human-readable match explanation
 */
function generateMatchExplanation(
  resume: Resume,
  job: Job,
  scores: MatchScore,
  matchedSkills: string[],
  missingSkills: string[]
): MatchExplanation {
  const strengths: string[] = [];
  const gaps: string[] = [];
  const recommendations: string[] = [];

  // Analyze skill match
  const skillMatchRatio = matchedSkills.length / Math.max(job.requiredSkills.length, 1);
  
  if (skillMatchRatio >= 0.8) {
    strengths.push(`Strong skill alignment with ${matchedSkills.length}/${job.requiredSkills.length} required skills`);
  } else if (skillMatchRatio >= 0.5) {
    strengths.push(`Good skill match with ${matchedSkills.length} relevant skills`);
    gaps.push(`Missing ${missingSkills.length} required skills`);
  } else {
    gaps.push(`Limited skill overlap (${matchedSkills.length}/${job.requiredSkills.length} required skills)`);
  }

  // Analyze experience
  const experienceYears = estimateExperienceYears(resume);
  if (scores.experienceScore >= 0.8) {
    strengths.push(`Experience level (${experienceYears.toFixed(1)} years) aligns well with ${job.experienceLevel} role`);
  } else if (scores.experienceScore < 0.5) {
    gaps.push(`Experience level may not match ${job.experienceLevel} requirements`);
  }

  // Analyze semantic match
  if (scores.semanticScore >= 0.7) {
    strengths.push('Strong semantic alignment between resume and job description');
  }

  // Generate recommendations
  if (missingSkills.length > 0 && missingSkills.length <= 5) {
    recommendations.push(`Consider gaining experience in: ${missingSkills.slice(0, 3).join(', ')}`);
  }

  if (scores.experienceScore < 0.7) {
    recommendations.push('Highlight relevant projects and coursework to strengthen experience match');
  }

  // Generate summary
  let summary = '';
  if (scores.overallScore >= 0.85) {
    summary = 'Excellent match! This candidate has strong qualifications for this role.';
  } else if (scores.overallScore >= 0.70) {
    summary = 'Good match with relevant skills and experience. Worth considering.';
  } else if (scores.overallScore >= 0.55) {
    summary = 'Fair match. Some qualifications align, but there are gaps to address.';
  } else {
    summary = 'Limited match. Significant skill or experience gaps exist.';
  }

  return {
    summary,
    strengths,
    gaps,
    recommendations,
    skillBreakdown: {
      matched: matchedSkills,
      missing: missingSkills,
      additional: resume.extractedSkills.filter(
        s => !job.requiredSkills.some(r => r.toLowerCase() === s.toLowerCase())
      ),
    },
  };
}

/**
 * Batch match multiple resumes to multiple jobs
 */
export function batchMatch(
  resumes: Resume[],
  jobs: Job[]
): Map<string, Map<string, MatchResult>> {
  logger.info('Starting batch matching', {
    resumeCount: resumes.length,
    jobCount: jobs.length,
    totalMatches: resumes.length * jobs.length,
  });

  const results = new Map<string, Map<string, MatchResult>>();

  for (const resume of resumes) {
    if (resume.status !== 'completed') continue;

    const jobMatches = new Map<string, MatchResult>();
    for (const job of jobs) {
      if (job.status !== 'active') continue;

      const match = matchResumeToJob(resume, job);
      jobMatches.set(job.id, match);
    }

    results.set(resume.id, jobMatches);
  }

  logger.info('Batch matching completed', {
    resumesProcessed: results.size,
  });

  return results;
}
